File: relay/common/per_second_pricing.go
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
package common

import (
	"fmt"
	"net/url"
	"strings"

	rootcommon "github.com/QuantumNous/new-api/common"
	"github.com/QuantumNous/new-api/setting/billing_setting"
	"github.com/gin-gonic/gin"
	"github.com/tidwall/gjson"
)

// ResolvePerSecondUnitPrice applies the first matching conditional price rule.
// Rules read fields from the original JSON, URL-encoded, or multipart request
// without changing the body sent upstream, then fall back to the normalized
// task request for protocol aliases.
func ResolvePerSecondUnitPrice(c *gin.Context, modelName string, defaultPrice float64) (float64, string, bool, error) {
	rules := billing_setting.GetPerSecondRules(modelName)
	if len(rules) == 0 {
		return defaultPrice, "", false, nil
	}

	var rawBody []byte
	if storage, err := rootcommon.GetBodyStorage(c); err == nil {
		rawBody, _ = storage.Bytes()
	}
	jsonBody := gjson.ValidBytes(rawBody)

	var formValues url.Values
	contentType := strings.ToLower(c.GetHeader("Content-Type"))
	switch {
	case strings.Contains(contentType, "multipart/form-data"):
		form, err := rootcommon.ParseMultipartFormReusable(c)
		if err != nil {
			return 0, "", true, fmt.Errorf("parse multipart per-second pricing fields: %w", err)
		}
		defer form.RemoveAll()
		formValues = url.Values(form.Value)
		for field, files := range form.File {
			if _, exists := formValues[field]; exists {
				continue
			}
			for _, file := range files {
				formValues[field] = append(formValues[field], file.Filename)
			}
		}
	case strings.Contains(contentType, "application/x-www-form-urlencoded"):
		values, err := url.ParseQuery(string(rawBody))
		if err != nil {
			return 0, "", true, fmt.Errorf("parse form per-second pricing fields: %w", err)
		}
		formValues = values
	}

	var normalizedBody []byte
	if request, err := GetTaskRequest(c); err == nil {
		normalizedBody, _ = rootcommon.Marshal(request)
	}

	price, ruleName, matched, err := billing_setting.MatchPerSecondRule(rules, func(path string) (any, bool) {
		for _, candidate := range perSecondRuleCandidatePaths(path) {
			if jsonBody {
				if result := gjson.GetBytes(rawBody, candidate); result.Exists() {
					return result.Value(), true
				}
			}
			if value, exists := perSecondFormValue(formValues, candidate); exists {
				return value, true
			}
			if result := gjson.GetBytes(normalizedBody, candidate); result.Exists() {
				return result.Value(), true
			}
		}
		return nil, false
	})
	if err != nil {
		return 0, "", true, err
	}
	if !matched {
		return defaultPrice, "", true, nil
	}
	return price, ruleName, true, nil
}

func perSecondFormValue(values url.Values, path string) (any, bool) {
	if len(values) == 0 {
		return nil, false
	}
	if fieldValues, exists := values[path]; exists && len(fieldValues) > 0 {
		return fieldValues[0], true
	}

	root, nestedPath, nested := strings.Cut(path, ".")
	if !nested {
		return nil, false
	}
	for _, value := range values[root] {
		if result := gjson.Get(value, nestedPath); result.Exists() {
			return result.Value(), true
		}
	}
	return nil, false
}

func perSecondRuleCandidatePaths(path string) []string {
	path = strings.TrimSpace(strings.TrimPrefix(strings.TrimSpace(path), "request."))
	switch strings.ToLower(path) {
	case "resolution":
		return []string{"resolution_name", "resolution", "metadata.resolution_name", "metadata.resolution", "size", "metadata.size"}
	case "quality":
		return []string{"quality", "metadata.quality"}
	case "mode":
		return []string{"mode", "metadata.mode"}
	default:
		return []string{path}
	}
}

File: web/src/features/system-settings/models/per-second-pricing.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
export type PerSecondOperator =
  | 'eq'
  | 'not_eq'
  | 'contains'
  | 'not_contains'
  | 'exists'
  | 'not_exists'
  | 'gt'
  | 'gte'
  | 'lt'
  | 'lte'

export type PerSecondConditionDraft = {
  id: string
  path: string
  operator: PerSecondOperator
  value: string
}

export type PerSecondRuleDraft = {
  id: string
  name: string
  price: string
  conditions: PerSecondConditionDraft[]
}

export type PerSecondRuleConfig = {
  name: string
  price: number
  conditions: Array<{
    path: string
    operator: PerSecondOperator
    value?: string
  }>
}

export const perSecondOperators: Array<{
  value: PerSecondOperator
  labelKey: string
}> = [
  { value: 'eq', labelKey: 'Equals' },
  { value: 'not_eq', labelKey: 'Does not equal' },
  { value: 'contains', labelKey: 'Contains' },
  { value: 'not_contains', labelKey: 'Does not contain' },
  { value: 'exists', labelKey: 'Exists' },
  { value: 'not_exists', labelKey: 'Does not exist' },
  { value: 'gt', labelKey: 'Greater than' },
  { value: 'gte', labelKey: 'Greater than or equal to' },
  { value: 'lt', labelKey: 'Less than' },
  { value: 'lte', labelKey: 'Less than or equal to' },
]

let perSecondDraftId = 0

function createPerSecondDraftId(type: 'rule' | 'condition'): string {
  perSecondDraftId += 1
  return `per-second-${type}-${perSecondDraftId}`
}

export function createPerSecondCondition(): PerSecondConditionDraft {
  return {
    id: createPerSecondDraftId('condition'),
    path: '',
    operator: 'eq',
    value: '',
  }
}

export function createPerSecondRule(index: number): PerSecondRuleDraft {
  return {
    id: createPerSecondDraftId('rule'),
    name: `Rule ${index + 1}`,
    price: '',
    conditions: [{ ...createPerSecondCondition(), path: 'resolution' }],
  }
}

export function normalizePerSecondRules(value: unknown): PerSecondRuleDraft[] {
  if (!Array.isArray(value)) return []
  return value.flatMap((candidate) => {
    if (!candidate || typeof candidate !== 'object') return []
    const rule = candidate as Partial<PerSecondRuleConfig>
    const conditions = Array.isArray(rule.conditions)
      ? rule.conditions.flatMap((condition) => {
          if (!condition || typeof condition !== 'object') return []
          const item = condition as Partial<
            PerSecondRuleConfig['conditions'][number]
          >
          const operator = perSecondOperators.some(
            (option) => option.value === item.operator
          )
            ? item.operator
            : 'eq'
          return [
            {
              id: createPerSecondDraftId('condition'),
              path: typeof item.path === 'string' ? item.path : '',
              operator: operator as PerSecondOperator,
              value: typeof item.value === 'string' ? item.value : '',
            },
          ]
        })
      : []
    return [
      {
        id: createPerSecondDraftId('rule'),
        name: typeof rule.name === 'string' ? rule.name : '',
        price:
          typeof rule.price === 'number' && Number.isFinite(rule.price)
            ? String(rule.price)
            : '',
        conditions,
      },
    ]
  })
}

export function validatePerSecondRules(
  rules: PerSecondRuleDraft[]
): string | null {
  for (const rule of rules) {
    if (!rule.name.trim()) return 'Each conditional price rule needs a name.'
    const price = Number(rule.price)
    if (rule.price === '' || !Number.isFinite(price) || price < 0) {
      return 'Each conditional price rule needs a valid non-negative price.'
    }
    if (rule.conditions.length === 0) {
      return 'Each conditional price rule needs at least one condition.'
    }
    for (const condition of rule.conditions) {
      if (!condition.path.trim()) {
        return 'Each condition needs a request field.'
      }
      if (
        condition.operator !== 'exists' &&
        condition.operator !== 'not_exists' &&
        !condition.value.trim()
      ) {
        return 'Each condition needs a comparison value.'
      }
    }
  }
  return null
}

export function serializePerSecondRules(
  rules: PerSecondRuleDraft[]
): PerSecondRuleConfig[] {
  return rules.map((rule) => ({
    name: rule.name.trim(),
    price: Number(rule.price),
    conditions: rule.conditions.map((condition) => ({
      path: condition.path.trim(),
      operator: condition.operator,
      ...(condition.operator === 'exists' || condition.operator === 'not_exists'
        ? {}
        : { value: condition.value.trim() }),
    })),
  }))
}
